src/utils/stark/starkCurve.ts

```diff
@@ -100,3 +109,58 @@ export function generateStarkPrivateKey(): string {
   const keyPair = starkEc.genKeyPair();
   return grindKey(keyPair.getPrivate(), starkEcOrder);
 }
+
+function getIntFromBits(
+  hex: string,
+  start: number,
+  end: number | undefined = undefined,
+): number {
+  const bin = encUtils.hexToBinary(hex);
+  const bits = bin.slice(start, end);
+  const int = encUtils.binaryToNumber(bits);
+  return int;
+}
+
+function getAccountPath(
+  layer: string,
+  application: string,
+  ethereumAddress: string,
+  index: string,
+): string {
+  const layerHash = hash.sha256().update(layer).digest('hex');
+  const applicationHash = hash.sha256().update(application).digest('hex');
+  const layerInt = getIntFromBits(layerHash, -31);
+  const applicationInt = getIntFromBits(applicationHash, -31);
+  const ethAddressInt1 = getIntFromBits(ethereumAddress, -31);
+  const ethAddressInt2 = getIntFromBits(ethereumAddress, -62, -31);
+  return `m/2645'/${layerInt}'/${applicationInt}'/${ethAddressInt1}'/${ethAddressInt2}'/${index}`;
+}
+
+function getKeyFromPath(seed: string, path: string): string {
+  const privateKey = hdkey
+    .fromMasterSeed(Buffer.from(seed.slice(2), 'hex')) // assuming seed is '0x...'
+    .derivePath(path)
+    .getWallet()
+    .getPrivateKey();
+  return grindKey(new BN(privateKey), starkEcOrder);
+}
+
+/**
+ * Generates a deterministic Stark private key from the provided signer.
+ * @returns the private key as a hex string
+ */
+export async function generateLegacyStarkPrivateKey(
+  signer: Signer,
+): Promise<string> {
+  const address = (await signer.getAddress()).toLowerCase();
+  const signature = await signer.signMessage(DEFAULT_SIGNATURE_MESSAGE);
+  const seed = splitSignature(signature).s;
+  const path = getAccountPath(
+    DEFAULT_ACCOUNT_LAYER,
+    DEFAULT_ACCOUNT_APPLICATION,
+    address,
+    DEFAULT_ACCOUNT_INDEX,
+  );
+  const key = getKeyFromPath(seed, path);
+  return key.padStart(64, '0');
```

Comment:
where does the padding requirement comes from? I'm asking because we dont do that on mobile and I dont remember it being required

Reply:
I just check it against the output of `github.com/immutable/generate-stark-key/` which uses v0.7.0 of core-sdk it outputs 64 chars for the test key and this current method was outputting 63 for the same.

Reply:
Actually the padding doesn't matter and can be removed it's just formatting.

Reply:
I don't mind it, as long as we don't do any type of silly comparison of the strings where there would be a mismatch, but I guess we wouldnt?

Reply:
That's right when we need a ec keypair, it is read by the crypto library as a hex so it doesn't care about the padded zeros.